deepmd/train/trainer.py

```diff
@@ -655,60 +805,94 @@ def valid_on_the_fly(self,
+                    for k in train_results[fitting_key].keys():
+                        print_str += prop_fmt % (k + '_trn')
         print_str += '   %8s\n' % 'lr'
         fp.write(print_str)
         fp.flush()
 
     @staticmethod
-    def print_on_training(fp, train_results, valid_results, cur_batch, cur_lr):
+    def print_on_training(fp, train_results, valid_results, cur_batch, cur_lr, multi_task_mode=False):
         print_str = ''
         print_str += "%7d" % cur_batch
-        if valid_results is not None:
-            prop_fmt = "   %11.2e %11.2e"
-            for k in valid_results.keys():
-                # assert k in train_results.keys()
-                print_str += prop_fmt % (valid_results[k], train_results[k])
+        if not multi_task_mode:
+            if valid_results is not None:
+                prop_fmt = "   %11.2e %11.2e"
+                for k in valid_results.keys():
+                    # assert k in train_results.keys()
+                    print_str += prop_fmt % (valid_results[k], train_results[k])
+            else:
+                prop_fmt = "   %11.2e"
+                for k in train_results.keys():
+                    print_str += prop_fmt % (train_results[k])
         else:
-            prop_fmt = "   %11.2e"
-            for k in train_results.keys():
-                print_str += prop_fmt % (train_results[k])
+            for fitting_key in train_results:
+                if valid_results[fitting_key] is not None:
+                    prop_fmt = "   %11.2e %11.2e"
+                    for k in valid_results[fitting_key].keys():
+                        # assert k in train_results[fitting_key].keys()
+                        print_str += prop_fmt % (valid_results[fitting_key][k], train_results[fitting_key][k])
+                else:
+                    prop_fmt = "   %11.2e"
+                    for k in train_results[fitting_key].keys():
+                        print_str += prop_fmt % (train_results[fitting_key][k])
         print_str += "   %8.1e\n" % cur_lr
         fp.write(print_str)
         fp.flush()
 
     def get_evaluation_results(self, batch_list):
-        if batch_list is None: return None
-        numb_batch = len(batch_list)
-
-        sum_results = {}    # sum of losses on all atoms
-        sum_natoms = 0
-        for i in range(numb_batch):
-            batch = batch_list[i]
-            natoms = batch["natoms_vec"]
-            feed_dict = self.get_feed_dict(batch, is_training=False)
-            results = self.loss.eval(self.sess, feed_dict, natoms)
-
-            for k, v in results.items():
-                if k == "natoms":
-                    sum_natoms += v
-                else:
-                    sum_results[k] = sum_results.get(k, 0.) + v * results["natoms"]
-        avg_results = {k: v / sum_natoms for k, v in sum_results.items() if not k == "natoms"}
+        def eval_single_list(single_batch_list, loss, sess, get_feed_dict_func, prefix=''):
```

Comment:
I am not sure if we need to introduce the `eval_single_list` in the scope `_build_training`.